src/Compilers/CSharp/Portable/Symbols/FileIdentifier.cs

```diff
@@ -4,38 +4,98 @@
 
 using System.Collections.Immutable;
 using System.Text;
+using System.Threading;
 using Microsoft.CodeAnalysis.CSharp.Symbols;
 using Microsoft.CodeAnalysis.Text;
 
 namespace Microsoft.CodeAnalysis.CSharp;
 
-internal readonly struct FileIdentifier
+internal sealed class FileIdentifier
 {
     private static readonly Encoding s_encoding = new UTF8Encoding(encoderShouldEmitUTF8Identifier: false, throwOnInvalidBytes: true);
 
-    public string? EncoderFallbackErrorMessage { get; init; }
-    public ImmutableArray<byte> FilePathChecksumOpt { get; init; }
-    public string DisplayFilePath { get; init; }
+    private string? _filePath;
 
-    public static FileIdentifier Create(SyntaxTree tree)
-        => Create(tree.FilePath);
+    private string? _encoderFallbackErrorMessage;
+    private string? _displayFilePath;
+    private ImmutableArray<byte> _filePathChecksumOpt;
 
-    public static FileIdentifier Create(string filePath)
+    private FileIdentifier(string filePath)
+    {
+        _filePath = filePath;
+    }
+
+    private FileIdentifier(ImmutableArray<byte> filePathChecksumOpt, string displayFilePath)
+    {
+        _filePathChecksumOpt = filePathChecksumOpt;
+        _displayFilePath = displayFilePath;
+    }
+
+    private void EnsureInitialized()
+    {
+        var filePath = _filePath;
+        if (filePath is not null)
+        {
+            string? encoderFallbackErrorMessage = null;
+            ImmutableArray<byte> hash = default;
+            try
+            {
+                var encodedFilePath = s_encoding.GetBytes(filePath);
+                using var hashAlgorithm = SourceHashAlgorithms.CreateDefaultInstance();
+                hash = hashAlgorithm.ComputeHash(encodedFilePath).ToImmutableArray();
+            }
+            catch (EncoderFallbackException ex)
+            {
+                encoderFallbackErrorMessage = ex.Message;
+            }
+
+            var displayFilePath = GeneratedNames.GetDisplayFilePath(filePath);
+
+            _encoderFallbackErrorMessage = encoderFallbackErrorMessage;
+            _displayFilePath = displayFilePath;
+
+            ImmutableInterlocked.InterlockedInitialize(ref _filePathChecksumOpt, hash);
+
+            Volatile.Write(ref _filePath, null);
```

Comment:
@RikkiGibson -- do you still have concerns?

Reply:
I just saw Cyrus use StrongBox in another PR for a similar concern. I think I would feel better about _filePathChecksumOpt being a StrongBox<byte>? instead of an ImmutableArray

Reply:
Cyrus's fix wasn't around usage of InterlockedInitialize, but was around concurrent read/write of a struct. With the change from commit 7 to commit 8, I no longer need to worry about the _filePathChecksumOpt member getting partially written during another thread reading it. I have no problem changing it back if you are convinced that wouldn't have been a problem.

Reply:
> I have no problem changing it back if you are convinced that wouldn't have been a problem.

It is a very common pattern that we use in symbols to rely on ```ImmutableInterlocked.InterlockedInitialize``` and simply read the array in another place. It feels like StrongBox is an overkill here, or all other places should be fixed up as well.

Reply:
I like the idea of putting all the data into a single reference type. I'll go with something like the following unless there is a good reason someone has to use Tuple directly instead:

private record FileIdentifierData(string? EncoderFallbackErrorMessage, string DisplayFilePath, ImmutableArray<byte> FilePathChecksumOpt);